v0.12 upgrading guide (#560)
membraneframework/membrane_core: guides/upgrading/v0.12.md

```diff
- spec = 
-   get_child(:source)
-   |> child(:sink, My.Sink)
- 
- actions = [spec: spec, notify_child: {:source, :start_streaming}]  
- 
- {actions, state}

+ spec = 
+   get_child(Child.ref(:source, group: :source_group))
+   |> child(:sink, My.Sink)
+
+ actions = [
+   spec: spec,
+   notify_child: {Child.ref(:source, group: :source_group), :start_streaming}
+ ]  
+
+ {actions, state}
```

Analogically, if a component is spawned inside a group, value of field `name` in it's callback context changes as well, and becames `Child.ref(child_name, group: group_name)`

```diff
- @impl true 
- def handle_init(%{name: :source} = _ctx, state) do
-   ...
- end

+ @impl true
+ def handle_init(%{name: Child.ref(:source, group: :source_group)} = _ctx, state) do
+   ...
+ end
```

For components spawned beyond groups nothing changes

## Remove `:playback` action

Now, membrane pipelines enter the playing playback by default and they don't have to return a `:playback` action to do it. 

```diff
- @impl true 
- def handle_setup(_ctx, state) do
- {[playback: :playing], state}
- end
```

> Instead of it, there is a new action introduced in `membrane_core` v0.12, `setup: :incomplete | :complete`. If you want to defer a moment when a component enters the playing playback, you can return `{:setup, :incomplete}` action from `handle_setup` callback. If you do that, a component will enter the playing playback only when you return `{:setup, :complete}` action from the another callback, eg. `handle_info`.

Reviewer: Check if grammarly reports more errors

```suggestion
Instead of it, there is a new action introduced in `membrane_core` v0.12, `setup: :incomplete | :complete`. If you want to defer a moment when a component enters the playing playback, you can return `{:setup, :incomplete}` action from `handle_setup` callback. If you do that, a component will enter the playing playback only when you return `{:setup, :complete}` action from another callback, eg. `handle_info`.
```

# Upgrading to v0.12

> Between v0.11 and v0.12 some breaking changes have occurred, so here comes the guide that will help you adjust your code to the new API. See the [changelog](https://github.com/membraneframework/membrane_core/releases/tag/v0.12.0) for details.

Reviewer 2: ```suggestion
Between v0.11 and v0.12 some breaking changes have occurred, so here comes the guide that will help you adjust your code to the new API. See the [release notes](https://github.com/membraneframework/membrane_core/releases/tag/v0.12.0) for details.
```
You haven't linked changelog :D

# Upgrading to v0.12

Between v0.11 and v0.12 some breaking changes have occurred, so here comes the guide that will help you adjust your code to the new API. See the [changelog](https://github.com/membraneframework/membrane_core/releases/tag/v0.12.0) for details.

## Deps upgrade

Update `membrane_core` to `v0.12`
```elixir 
defp deps do
   [
     {:membrane_core, "~> 0.12.0"},
     ...
   ]
end
```

## Implement `handle_child_pad_removed/4` callback in bins and pipelines, if it is needed

Now, if bin removes its pad (eg. by removing an element linked to the bin's inner pad), bin's parent has to have implemented proper `handle_child_pad_removed/4` callback, to handle it. If there is no such a callback, default behaviour is to raise an error. 

```elixir
@impl true
def handle_child_pad_removed(:rtp, Pad.ref(:rtp_input, _ssrc), _ctx, state) do
  # ...
end
```

## Use `Child.ref/2`, if you want to refer a child spawned inside a group

Until now, name of a child spawned in a spec like this: 

> ```elixir
> {child(:source, %My.Source{options: options}), crash_group: {:source_group, :temporary}}
> ```

Reviewer: ````suggestion
```diff
- {child(:source, %My.Source{options: options}), crash_group: {:source_group, :temporary}}
```
````

and add the 0.12 version

Author: What do you mean? This is 0.12 version

would be just `:source`. 

Since `membrane_core` v0.12, if you spawn a child inside a group (this includes crash group as well), child name will also contain name of the group. In the case above, name of the `My.Source` will be `Child.ref(:source, group: :source_group)`. If you want to send a notification to it or link it with the another child, you can do it as follows:

> ```elixir
> spec = 
>   get_child(Child.ref(:source, group: :source_group))
>   |> child(:sink, My.Sink)
> 
> actions = [
>   spec: spec,
>   notify_child: {Child.ref(:source, group: :source_group), :start_streaming}
> ]  
> 
> {actions, state}
> ```

Reviewer: ````suggestion
```diff
+ spec = 
+   get_child(Child.ref(:source, group: :source_group))
+   |> child(:sink, My.Sink)
+ 
+ actions = [
+   spec: spec,
+   notify_child: {Child.ref(:source, group: :source_group), :start_streaming}
+ ]
+ 
+ {actions, state}
```
````

and add the pre-0.12 version like in the other examples

## Adjust to change of value of field `name` in callbacks context

> 

Reviewer: I'd merge it into the previous point

> If a component is spawned inside a group, value of field `name` in it's callback context changes as well, and becames `Child.ref(child_name, group: group_name)`   

Reviewer: add an example

Since `membrane_core` v0.12, if you spawn a child inside a group (this includes crash group as well), child name will also contain name of the group. In the case above, name of the `My.Source` will be `Child.ref(:source, group: :source_group)`. If you want to send a notification to it or link it with the another child, you can do it as follows:

```diff
- spec = 
-   get_child(:source)
-   |> child(:sink, My.Sink)
- 
- actions = [spec: spec, notify_child: {:source, :start_streaming}]  
- 
- {actions, state}

+ spec = 
+   get_child(Child.ref(:source, group: :source_group))
+   |> child(:sink, My.Sink)
+
+ actions = [
+   spec: spec,
+   notify_child: {Child.ref(:source, group: :source_group), :start_streaming}
+ ]  
+
+ {actions, state}
```

Analogically, if a component is spawned inside a group, value of field `name` in it's callback context changes as well, and becames `Child.ref(child_name, group: group_name)`

```diff
- @impl true 
- def handle_init(%{name: :source} = _ctx, state) do
-   ...
- end

+ @impl true
+ def handle_init(%{name: Child.ref(:source, group: :source_group)} = _ctx, state) do
+   ...
+ end

> ```

Reviewer: there's no way to have a child in handle_init's context, let's change callback to handle_info
